File: Pow(x,n).cpp
class Solution {
public:
    double power(double x, int n) {
        if (n == 0) return 1;
        else if (n == 1) return x;
        
        double res = power(x, n/2);
        res *= res;
        if (n & 0x1) {
            res *= x;
        }
        return res;
    }
    double pow(double x, int n) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        
        bool sign = n >= 0;
        n = abs(n);
        
        double res = power(x, n);
        return sign ? res : 1/res;
    }
};


class Solution {
public:
    double power(double x, unsigned n) {
        double res = 1;
        while (n > 0) {
            if ((n&0x1) == 1) {
                res *= x;
            }
            x *= x;
            n >>= 1;
        }
        return res;
    }
    double pow(double x, int n) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (n == 0) {
            return 1;
        }
        
        bool sign = n > 0;
        n = abs(n);
        double res = power(x, n);
        return sign ? res : 1/res;
    }
};

File: Pow(x, n).cs
public class Solution {
    public double MyPow(double x, int n) {
        var sign = true;
        if (n < 0) {
            if (n == int.MinValue) {
                if (x < 0) {
                    x = -x;
                }
                x = 1 / x;
                return x * Pow(x, -(n + 1));
            } else {
                n = -n;
            }
            x = 1 / x;
        }
        
        return Pow(x, n);
    }
    
    private double Pow(double x, int n) {
        if (n == 0) {
            return 1;
        } else if (n == 1) {
            return x;
        }
        
        double power = Pow(x, n / 2);
        power *= power;
        if (n % 2 == 1) {
            power *= x;
        }
        
        return power;
    }
}
